### PR TITLE
Control flow overhaul of Prover repl and addition of `Undo`  prover command

### PR DESCRIPTION
**Pull request by njohnwalker**
improvements:
- Reimplemented the prover repl to utilize monad transformers for a more painless integration of stateful computation on matching logic proofs.
- added undo functionality

Further work:
- move functionality (to reorder proof claims) should be a minor addition.
- Testing.  I have not added tests for the user interface or the undo functionality.
- updating the `prover_repl.md`

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

File: src/main/haskell/kore/app/prover/Main.hs [hunk 1]
```diff
@@ -1,29 +1,44 @@
...
+import GlobalMain
+       ( defaultMainGlobal )
 
 -- TODO: still needed?
 parseName :: Parser String
 parseName = takeWhile1P Nothing isAlphaNum <* space
 
-pCommand' :: Parser (Command String (MLRule (SymbolOrAlias Meta) (Variable Meta) CommonMetaPattern) CommonMetaPattern)
-pCommand' = parseCommand parseName parseFormula parseRule
+pCommand :: Parser (Command String (MLRule (SymbolOrAlias Meta) (Variable Meta) CommonMetaPattern) CommonMetaPattern)
```
Comment (ttuegel) [inline src/main/haskell/kore/app/prover/Main.hs]: This monster type should have a synonym. Could I suggest this?
```.hs
type CommonRule = MLRule (SymbolOrAlias Meta) (Variable Meta) CommonMetaPattern
```

File: src/main/haskell/kore/app/prover/Main.hs [hunk 2]
```diff
@@ -32,11 +47,41 @@ pCommand' = parseCommand parseName parseFormula parseRule
                                parseName
 
 proveCommand
-    :: IO (ProverState
-            String
-            (MLRule (SymbolOrAlias Meta) (Variable Meta) CommonMetaPattern)
-            CommonMetaPattern)
-proveCommand = runProver dummyFormulaVerifier pCommand' (ProverState emptyProof)
+  :: (Reifies sig ValidatedSignature)
```
Comment (ttuegel) [inline src/main/haskell/kore/app/prover/Main.hs]: Where is this `Reifies` constraint used?

File: src/main/haskell/kore/app/prover/Main.hs [hunk 3]
```diff
@@ -32,11 +47,41 @@ pCommand' = parseCommand parseName parseFormula parseRule
...
 main :: IO ()
-main = proveCommand >> return ()
+main =
+  defaultMainGlobal
+  >> ( case testSigValid of
+         Nothing -> putStrLn $ show testSigValid
+         Just validSig ->
+           reifySignature
+           validSig
+           (\proxy -> proveCommand proxy
```
Comment (ttuegel) [inline src/main/haskell/kore/app/prover/Main.hs]:
```.hs
(\proxy -> proveCommand proxy >> return ())
===
(void . proveCommand)
```

File: src/main/haskell/kore/app/prover/Main.hs (hunk 2 above)
Comment (njohnwalker) [reply src/main/haskell/kore/app/prover/Main.hs]: This is some leftover stuff from brandon's implementation that I didn't understand so I left it for now

File: src/main/haskell/kore/app/prover/Main.hs (hunk 3 above)
Comment (njohnwalker) [reply src/main/haskell/kore/app/prover/Main.hs]: see: https://github.com/kframework/kore/pull/175/review_comment/create#discussion_r208700463

**Pull request merged**